internal/v2/security.go

```diff
@@ -0,0 +1,119 @@
+package v2
+
+import (
+	"bytes"
+	"context"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"sort"
+	"strings"
+	"time"
+)
+
+// SecurityProviderExoscaleV2 represents an Exoscale API V2 security provider.
+type SecurityProviderExoscaleV2 struct {
+	// ReqExpire represents the request expiration duration.
+	ReqExpire time.Duration
+
+	apiKey    string
+	apiSecret string
+}
+
+// NewSecurityProviderExoscaleV2 returns a new Exoscale API V2 security provider to sign API requests using the
+// specified API key/secret.
+func NewSecurityProviderExoscaleV2(apiKey, apiSecret string) (*SecurityProviderExoscaleV2, error) {
+	var provider = SecurityProviderExoscaleV2{
+		ReqExpire: 10 * time.Minute,
+	}
+
+	if provider.apiKey = apiKey; provider.apiKey == "" {
+		return nil, errors.New("missing API key")
+	}
+
+	if provider.apiSecret = apiSecret; provider.apiSecret == "" {
+		return nil, errors.New("missing API secret")
+	}
+
+	return &provider, nil
+}
+
+// Intercept is an HTTP middleware that intercepts and signs client requests before sending them to the API
+// endpoint.
+func (s *SecurityProviderExoscaleV2) Intercept(_ context.Context, req *http.Request) error {
+	return s.signRequest(req, time.Now().UTC().Add(s.ReqExpire))
+}
+
+func (s *SecurityProviderExoscaleV2) signRequest(req *http.Request, expiration time.Time) error {
+	var buf bytes.Buffer
+
+	authHeader := fmt.Sprintf("EXO2-HMAC-SHA256 credential=%s", s.apiKey)
```

Comment:
Instead of concatenating strings in `authHeader` how about doing something like `authHeaderParts = []string{}` then you append into this array whenever needed. And once you are done with it you simple run `strings.Join(authHeaderParts, ',')` which will avoid any mistake regarding the `,` separator ?

Reply:
That's has also been suggested in https://github.com/exoscale/requests-exoscale-auth/pull/2 so you're right: I should use the same method here.

Reply:
Changed in f0d8769.

Reply:
You have applied the part thing method for `sigParts` but not for `authHeader` whereas it seems it gets also concatenated at some point. It there a reason for this ?

Reply:
Nope that was an oversight and you're right, I'll get to it right away 😳

Reply:
Fixed in 289d51a.